src/Manager/UndoRedoManager.ts

```diff
@@ -0,0 +1,114 @@
+import ActionType from '../Action/ActionType';
+import Shape from '../Shapes/Shape';
+import type ShapeManager from './ShapeManager';
+
+type UndoOrRedo = ActionType.undo | ActionType.redo;
+
+class UndoRedoManager {
+  private shapes: Array<Shape> = [];
+  private undoShapes: Array<Shape> = [];
+  private shapeManager: ShapeManager;
+
+  constructor(shapeManager: ShapeManager) {
+    this.shapeManager = shapeManager;
+  }
+
+  public addShape(shape: Shape): void {
+    this.shapes.push(shape);
+  }
+
+  public resetCreatedRedo(): void {
+    this.undoShapes = this.undoShapes.filter((shape) => !shape.created);
+  }
+
+  public getLastShape(): Shape | undefined {
+    if (this.shapes.length === 0) return undefined;
+    return this.shapes[this.shapes.length - 1];
+  }
+
+  public undo(): void {
+    this.shapeManager.drawFinish();
+    const shape: Shape | undefined = UndoRedoManager.popLastCreated(this.shapes);
+    if (shape !== undefined) {
+      this.emit(ActionType.undo, shape);
+      this.applyUndo(shape);
+    }
+  }
+
+  public externalUndo(id: string): void {
+    const shape: Shape = this.getAndRemove(this.shapes, id);
+    this.shapeManager.animationQueue.add(() => {
+      this.applyUndo(shape);
+      return Promise.resolve();
+    });
+  }
+
+  public redo(): void {
+    this.shapeManager.drawFinish();
+    const shape: Shape | undefined = UndoRedoManager.popLastCreated(this.undoShapes);
+    if (shape !== undefined) {
+      this.emit(ActionType.redo, shape);
+      this.applyRedo(shape);
+    }
+  }
+
+  public externalRedo(id: string): void {
+    const shape: Shape = this.getAndRemove(this.undoShapes, id);
+    this.shapeManager.animationQueue.add(() => {
+      this.applyRedo(shape);
+      return Promise.resolve();
+    });
+  }
+
+  // /!\ Mutating function
+  private static popLastCreated(shapes: Array<Shape>): Shape | undefined {
+    for (let i = shapes.length - 1; i >= 0; i -= 1) {
+      if (shapes[i].created) {
+        const shape = shapes[i];
+        shapes.splice(i, 1);
+        return shape;
+      }
+    }
+    return undefined;
+  }
+
+  // /!\ Mutating function
+  private getAndRemove(shapes: Array<Shape>, id: string) {
+    for (let i = shapes.length - 1; i >= 0; i -= 1) {
+      if (shapes[i].id === id) {
```

Comment:
Why not use find?

Reply:
Sounds good champ!
![image](https://user-images.githubusercontent.com/22148893/141197181-31644d6c-be1b-4b8e-8ac6-69fed43ac779.png)